tools/projmgr/src/ProjMgrYamlEmitter.cpp

```diff
@@ -71,6 +71,121 @@ class ProjMgrYamlCbuildIdx : public ProjMgrYamlBase {
+  for (auto& resolvedItem : csolution.cbuildPack.packs) {
+    ModelItem modelItem;
+    ProjMgrUtils::ConvertToPackInfo(resolvedItem.pack, modelItem.info);
+    modelItem.resolvedPack = resolvedItem;
+    ProjMgrUtils::PushBackUniquely(modelItem.resolvedPack.selectedBy, resolvedItem.pack);
+    model[resolvedItem.pack] = modelItem;
+  }
+
+  // Stage 2: Process packs that are required by used components
+  for (const auto& context : processedContexts) {
+    for (const auto& [packId, _] : context->packages) {
+// TODO: Skip if project local packs
+      if (model.find(packId) == model.end()) {
+        // Add pack
+        ModelItem modelItem;
+        ProjMgrUtils::ConvertToPackInfo(packId, modelItem.info);
+        modelItem.resolvedPack.pack = packId;
+        modelItem.resolvedPack.selectedBy.push_back(packId);
+        model[packId] = modelItem;
+      }
+    }
+  }
+
+  // Stage 3: Add all packs that do not contain wildcard and also not used by any component
+  for (const auto& context : processedContexts) {
+    for (const auto& packItem : context->packRequirements) {
+      // Skip project local packs
+      if (!packItem.path.empty()) {
+        continue;
+      }
+
+      const auto& reqInfo = packItem.pack;
+
+      // Skip wildcard packs
+      if (reqInfo.name.empty() || WildCards::IsWildcardPattern(reqInfo.name)) {
+        continue;
+      }
+
+      const string packId = RtePackage::ComposePackageID(reqInfo.vendor, reqInfo.name, reqInfo.version);
+      if (model.find(packId) == model.end()) {
+        // Pack does not have a version range, so add it
+        ModelItem modelItem;
+        ProjMgrUtils::ConvertToPackInfo(packId, modelItem.info);
+        modelItem.resolvedPack.pack = packId;
+        modelItem.resolvedPack.selectedBy.push_back(packId);
+        model[packId] = modelItem;
+      }
+    }
+  }
+
+  // Stage 4: Add all original expressions that was eagerly replaced
+  for (const auto& context : processedContexts) {
+    for (const auto& [key, val] : context->eagerlyReplacedPackIdMap) {
+      if (model.find(val) != model.end()) {
+        ProjMgrUtils::PushBackUniquely(model[val].resolvedPack.selectedBy, key);
+      }
+    }
+  }
+
+  // Stage 5: Process all wildcard patterns from user and add to selected-by list
+  for (const auto& context : processedContexts) {
+    for (const auto& packItem : context->packRequirements) {
+      // Skip project local packs
+      if (!packItem.path.empty()) {
+        continue;
+      }
+
+      const PackInfo& reqInfo = packItem.pack;
+      if (reqInfo.name.empty() || WildCards::IsWildcardPattern(reqInfo.name)) {
+        const string packId = RtePackage::ComposePackageID(reqInfo.vendor, reqInfo.name, reqInfo.version);
+
+        for (auto& [key, val] : model) {
+          if (ProjMgrUtils::IsMatchingPackInfo(val.info, reqInfo)) {
+            ProjMgrUtils::PushBackUniquely(val.resolvedPack.selectedBy, packId);
+          }
+        }
+      }
+    }
```

Comment:
For consistency and avoiding duplication, if possible I would recommend keeping the pack resolution in a common place, currently I think it’s done mostly in `ProjMgrWorker::GetRequiredPdscFiles`, and to store the information needed by the `selected-by` node close to the relevant variables in the pack resolution chain (`context.packRequirements`, `context.pdscFiles`, `m_loadedPacks`), eventually simplifying the emitter code here and also covering local packs that are commented as TODO.

Reply:
Project local packs are not included in the `cbuild-pack.yml` file.

The loading and the persisting of the `cbuild-pack.yml` content needs to be done before respectively after the RTE model has decided what packs that are to be used. See the diagram in the [documentation](https://github.com/slhultgren/cmsis-toolbox/blob/pr/shu-cbuild-pack-documentation/docs/YML-CBuild-Format.md#pack-locking) for details.